beacon-chain/forkchoice/doubly-linked-tree/store.go

```diff
@@ -99,17 +111,22 @@ func (s *Store) insert(ctx context.Context,
 		}
 	}
 
-	s.nodeByPayload[payloadHash] = n
-	s.nodeByRoot[root] = n
 	if parent == nil {
 		if s.treeRootNode == nil {
 			s.treeRootNode = n
 			s.headNode = n
 			s.highestReceivedNode = n
-		} else {
+		} else if s.treeRootNode.root != n.root {
 			return n, errInvalidParentRoot
 		}
-	} else {
+	}
+	if !rootPresent {
+		s.nodeByRoot[root] = n
+	}
+	if !hashPresent {
+		s.nodeByPayload[payloadHash] = n
+	}
```

Comment:
We update them both unconditionally which may cause trouble on collisions or on `payloadHash = [32]byte{}`.

Reply:
Is it worth checking both `!hashPresent ` and `payloadHash != [32]byte{}`? I guess it's fine as long as this work:
```
		if payloadHash == [32]byte{} {
			return n, nil
		}
```

Reply:
In general `[32]byte{}` should always return `hashPresent`